### DISCUSSION TITLE
BLE scanner

### DISCUSSION THREAD

**Discussion by nagyrobi** [HASwitchPlate/openHASP#85] [Q&A]
Somehow I was not aware that ESP32 has BLE capability, thank you for [pointing that out](https://github.com/fvanroie/hasp-lvgl/pull/82#issuecomment-792346647).

That gave me the idea that we could very simply implement a presence detection in these devices using a simple BLE scanner which would drop in the list of the detected BLE tags in the air to MQTT. For example having a plate in each room, additionally to the controls it offers, provides a great way to inform the automation system about persons being there or not.

I've created a new branch for this: https://github.com/nagyrobi/hasp-lvgl/tree/0.4.0-ble
Added a single file named `ble.cpp` which contains a simple BLE scanner (inspired from [moononournation](https://github.com/moononournation/Arduino_BLE_Scanner)).

Can you please point me in the right direction on how to integrate this in the main loop and publish the data to MQTT?

I plan to make this user-configurable through user_config_override so only the ones who require it would have it.

**Comment by fvanroie** [comment]
You would need to implement a bleSetup() and bleLoop() function. The bleLoop() function cannot be blocking because it would freeze the GUI. The .ino example scans for [30 seconds](https://github.com/nagyrobi/hasp-lvgl/blob/88c6e7408c33b2549851b4e5c7a67182f032b843/src/ble.cpp#L42), which would not be possible in combination with hasp-lvgl. You need to make that scan asynchronous.

BLE is very resource hungry. When we implement BLE it will probably be [NimBLE ](https://www.arduino.cc/reference/en/libraries/nimble-arduino/) based. This is a 3rd party BLE library that has a much smaller footprint then the Espressif one. NimBLE has an [continuous scanning example](https://github.com/h2zero/NimBLE-Arduino/blob/master/examples/NimBLE_Scan_Continuous/NimBLE_Scan_Continuous.ino). Instead of using a delay the loop would go in a bleEverySecond() or bleEvery5Seconds() function.

I haven't tried this myself. I think most phones implement an anti-tracking token, so I'm not sure if you can track an individual phone like this.

**Reply by dgomes** [reply]
Here's an idea: only scan when the panel is off (backlight off) my panel is off 99% of the time

**Reply by fvanroie** [reply]
If you can't instantly break-out of the 30-second scan interval, that still would be a very bad UX.

**Reply by dgomes** [reply]
Here's a different take: have the plate send beacons, used by the phone to detect where it is

**Reply by nagyrobi** [reply]
Good idea

**Comment by nagyrobi** [comment]
@fvanroie check this: https://github.com/jptrsn/ESP32-mqtt-room
Do you think it could be combined with openHASP?

**Reply by fvanroie** [reply]
It looks like it could be combined as a my_custom.cpp

**Reply by nagyrobi** [reply]
There's also https://github.com/HeimdallMidgard/ESP-32-BLE-Scanner
Wonder which one is more accurate or more easy to implement with openHASP.

Best part of both is that they feed straight into HA's mqtt_room component... no need for further CC development.

**Reply by fvanroie** [reply]
Probably a combination of both 😃 
- ESP32-mqtt-room uses a separate task so it won't block the UI
- ESP-32-BLE-Scanner uses NimBLE library which is more compact and efficient

**Reply by nagyrobi** [reply]
Worth looking at https://esphome.io/components/binary_sensor/ble_presence.html too

**Reply by mjone4** [reply]
Probably in line with this is Bermuda which uses the ble with esphome… would be a great feature

https://github.com/agittins/bermuda

**Reply by fvanroie** [reply]
I have very little BLE knowledge, so this isn't something I can implement. If you have the time and knowledge you can try using the custom code hooks.

But I think it's easier to just setup serperate ESP devices for BLE scanners instead of trying to cramp everything into one device.
